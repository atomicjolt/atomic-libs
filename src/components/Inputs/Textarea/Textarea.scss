@use "../../mixins";

.aje-input textarea {
  --textarea-height-sm: 80px;
  --textarea-height-md: 140px;
  --textarea-height-lg: 200px;
}
