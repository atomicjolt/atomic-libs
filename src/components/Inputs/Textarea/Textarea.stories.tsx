import { ComponentMeta, ComponentStory } from "@storybook/react";
import Textarea, { Props } from ".";

export default {
title: "Inputs/Textarea",
component: Textarea
} as ComponentMeta<typeof Textarea>

const Template: ComponentStory<typeof Textarea> = (args: Props) => (
  <Textarea {...args} />
);

export const Default = Template.bind({});
Default.args = {
  label: "Textarea label",
}

// Additional stories go here