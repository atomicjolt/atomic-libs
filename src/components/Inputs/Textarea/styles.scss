@use "../../defines" as *;

.aj-input {
  textarea {
    --textarea-height-sm: 80px;
    --textarea-height-md: 140px;
    --textarea-height-lg: 200px;

    border: 1px solid var(--neutral600);
    border-radius: 5px;
    min-height: 200px;
    width: 100%;
    padding: 8px 12px;
    @include regular;
    font-size: 1.6rem;
    color: var(--text-clr);
    resize: none;

    &:focus {
      border-color: var(--outline-clr-primary);
      outline: var(--input-outline);
    }

    &:read-only {
      border: none;
      background-color: var(--neutral100);
    }
  }

  &.can-resize textarea {
    resize: both;
  }
}
