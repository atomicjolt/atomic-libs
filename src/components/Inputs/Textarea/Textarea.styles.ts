import styled from "styled-components";
import mixins from "../../../styles/mixins";
import { ComponentWrapper } from "../../../styles/utils";

export const TextAreaWrapper = styled(ComponentWrapper)`
  ${mixins.Sizing}
  &.has-error {
    --input-border-color: var(--error-clr);
  }

  &.can-resize-x textarea {
    resize: horizontal;
  }

  &.can-resize-y textarea {
    resize: vertical;
  }

  &.can-resize textarea {
    resize: both;
  }
`;

export const StyledTextArea = styled.textarea`
  ${mixins.Regular}
  border: var(--input-border);
  border-radius: var(--input-border-radius);
  min-height: 80px;
  width: 100%;
  padding: calc(var(--input-font-size) / 2) var(--input-padding-horiz);
  font-size: var(--input-font-size);
  color: var(--input-text-clr);
  background-color: var(--input-bg-clr);
  resize: none;

  &:read-only {
    border-color: var(--neutral100);
    --input-bg-clr: var(--neutral100);
  }
  &:focus {
    box-shadow: 0 0 0 1px var(--input-border-color);
    outline: var(--input-outline);
    outline-color: var(--input-border-color);
  }
`;
