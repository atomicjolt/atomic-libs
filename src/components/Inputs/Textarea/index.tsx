import React from 'react'
import '../../general.scss'
import '../common.scss'
import './styles.scss'

export interface Props {
  /** Must include a label. Labels are always Sentence case. */
  label: string;
  /** Error text should be descriptive and explicit in meaning. */
  error?: string;
  /** For additional information (ex. date format mm/dd/yy) */
  message?: string;
  /** Placeholders aren't common, you should use the message instead for most placeholders so the user doesn't lose the information after entering text in the input. */
  placeholder?: string;
  /** The input size should reflect the expected size of its content. */
  height?: 'small' | 'medium' | 'large';
  resize?: boolean;
  value?: string;
  readonly?: boolean;
  disabled?: boolean;
  required?: boolean;
}

/** Textarea Component */
export default function Textarea({ 
    height = 'small', 
    resize = true,
    label, 
    error, 
    message, 
    placeholder, 
    value,
    readonly = false,
    disabled = false,
    required = false,
  }: Props) {

  const inputID = 'textarea';
  const errorID = 'errorText';
  /* Add a space before the added class rather than inside the className attr on the tag. Looks cleaner. */
  let errorClass = error ? ' has-error' : '';
  let disabledClass = disabled ? ' is-disabled' : '';
  let resizeClass = resize ? ' can-resize' : '';

  return (
    <div className={`aj-input is-${height}${resizeClass}${errorClass}${disabledClass}`}>
      <label className="aj-label" htmlFor={inputID}>
        {label}
        {message ?
          <p className="aj-label--message">{message}</p>
          : null
        }
      </label>
      <textarea
        id={inputID}
        aria-describedby={error ? errorID : ""}
        placeholder={placeholder}
        value={value}
        readOnly={readonly}
        disabled={disabled}
        required={required}
      />
      {error ?
        <p id={errorID} className="aj-label--error">{error}</p>
        : null
      }
    </div>
  )
}
