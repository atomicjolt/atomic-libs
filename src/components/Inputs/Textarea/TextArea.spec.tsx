import React from "react";
import { render, screen } from "@testing-library/react";
import Textarea from ".";

it("matches snapshot", () => {
  const result = render(
    <Textarea value="" onChange={() => {}} label="label" />
  );
  expect(result.asFragment()).toMatchSnapshot();
});
