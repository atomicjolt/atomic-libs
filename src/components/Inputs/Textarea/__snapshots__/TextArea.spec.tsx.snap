// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe sc-jSUZER  cRkWsK aje-input can-resize aje-input can-resize is-small"
  >
    <label
      class="sc-dkrFOg hmuQfz aje-label aje-label aje-label"
      for="TextArea-:r0:-textarea"
    >
      label
    </label>
    <textarea
      aria-describedby=""
      class="sc-gKPRtg rvtqf"
      id="TextArea-:r0:-textarea"
    />
  </div>
</DocumentFragment>
`;
